File: scripts/add_personal_files.py
#!/usr/bin/env python3
"""PERSONAL MODIFICATION: register fork-added Swift files into the Xcode project.

The xcodeproj Ruby gem corrupts this project on save (a SwiftPM product,
MLXVLM, trips its serializer), so we hand-edit project.pbxproj with targeted,
verified insertions that mirror the project's AA/BB id convention.
Idempotent: refuses to run if the markers are already present.
"""
import sys

PBX = "TypeWhisper.xcodeproj/project.pbxproj"

APP_FILES = [
    "FuzzyTermMatcher.swift",
    "DictionaryService+FuzzyTerms.swift",
    "DictionarySyncModels.swift",
    "DictionarySyncMerge.swift",
    "DictionarySyncService.swift",
    "PersonalSettingsSection.swift",
    "ReviewLearning.swift",
    "TranscriptReviewView.swift",
    "TranscriptReviewController.swift",
]
TEST_FILES = [
    "FuzzyTermMatcherTests.swift",
    "DictionarySyncMergeTests.swift",
    "ReviewLearningTests.swift",
]

def gid(prefix, n):
    return prefix + str(n).zfill(20)  # 22-char id, matches existing convention

def main():
    with open(PBX, "r") as f:
        text = f.read()

    if "Personal/FuzzyTermMatcher.swift" in text:
        print("Already registered; nothing to do.")
        return 0

    build_lines = []      # PBXBuildFile section
    fileref_lines = []     # PBXFileReference section
    app_group_children = []
    test_group_children = []
    app_sources = []
    test_sources = []

    def emit(basename, relpath, build_id, ref_id, is_test):
        build_lines.append(
            f'\t\t{build_id} /* {basename} in Sources */ = '
            f'{{isa = PBXBuildFile; fileRef = {ref_id} /* {basename} */; }};'
        )
        fileref_lines.append(
            f'\t\t{ref_id} /* {basename} */ = {{isa = PBXFileReference; '
            f'lastKnownFileType = sourcecode.swift; path = "{relpath}"; '
            f'sourceTree = SOURCE_ROOT; }};'
        )
        child = f'\t\t\t\t{ref_id} /* {basename} */,'
        src = f'\t\t\t\t{build_id} /* {basename} in Sources */,'
        if is_test:
            test_group_children.append(child)
            test_sources.append(src)
        else:
            app_group_children.append(child)
            app_sources.append(src)

    n = 9001
    for name in APP_FILES:
        emit(name, f"TypeWhisper/Personal/{name}", gid("AA", n), gid("BB", n), False)
        n += 1
    n = 9101
    for name in TEST_FILES:
        emit(name, f"TypeWhisperTests/{name}", gid("AA", n), gid("BB", n), True)
        n += 1

    # Anchored, single-occurrence replacements.
    edits = [
        ("/* End PBXBuildFile section */",
         "\n".join(build_lines) + "\n/* End PBXBuildFile section */"),
        ("/* End PBXFileReference section */",
         "\n".join(fileref_lines) + "\n/* End PBXFileReference section */"),
        ("\t\t\t\tBB00000000000000000079 /* DictionaryService.swift */,",
         "\t\t\t\tBB00000000000000000079 /* DictionaryService.swift */,\n" + "\n".join(app_group_children)),
        ("\t\t\t\tA8D4F2B6E1C3097B5F8A4E62 /* DictionaryExporterTests.swift */,",
         "\t\t\t\tA8D4F2B6E1C3097B5F8A4E62 /* DictionaryExporterTests.swift */,\n" + "\n".join(test_group_children)),
        ("\t\t\t\tAA00000000000000000079 /* DictionaryService.swift in Sources */,",
         "\t\t\t\tAA00000000000000000079 /* DictionaryService.swift in Sources */,\n" + "\n".join(app_sources)),
        ("\t\t\t\tC7A3E9F1D5B2084A6E9C3D71 /* DictionaryExporterTests.swift in Sources */,",
         "\t\t\t\tC7A3E9F1D5B2084A6E9C3D71 /* DictionaryExporterTests.swift in Sources */,\n" + "\n".join(test_sources)),
    ]

    for anchor, replacement in edits:
        count = text.count(anchor)
        if count != 1:
            print(f"ERROR: anchor found {count} times (expected 1): {anchor[:60]!r}")
            return 1
        text = text.replace(anchor, replacement, 1)

    with open(PBX, "w") as f:
        f.write(text)
    print(f"OK: registered {len(APP_FILES)} app + {len(TEST_FILES)} test files.")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: scripts/add_personal_files.rb
#!/usr/bin/env ruby
# PERSONAL MODIFICATION: registers fork-added source files into the Xcode
# project (the project uses explicit file references, not synchronized groups).
# Idempotent — running it twice will not create duplicate references.

require "xcodeproj"

# Workaround: this project references a SwiftPM product (MLXVLM) in build
# phases in a way that trips xcodeproj's annotation during save. The annotation
# is only a human-readable comment, so tolerate failures.
module Xcodeproj
  class Project
    module Object
      class PBXBuildFile
        def ascii_plist_annotation
          " /* #{display_name} */ "
        rescue StandardError
          " "
        end
      end
    end
  end
end

project_path = "TypeWhisper.xcodeproj"
project = Xcodeproj::Project.open(project_path)

main_target = project.targets.find { |t| t.name == "TypeWhisper" }
test_target = project.targets.find { |t| t.name == "TypeWhisperTests" }
abort("Missing targets") unless main_target && test_target

# Files: [relative_path_on_disk, target, group_path_segments]
app_files = [
  "TypeWhisper/Personal/FuzzyTermMatcher.swift",
  "TypeWhisper/Personal/DictionaryService+FuzzyTerms.swift",
  "TypeWhisper/Personal/DictionarySyncModels.swift",
  "TypeWhisper/Personal/DictionarySyncMerge.swift",
  "TypeWhisper/Personal/DictionarySyncService.swift",
  "TypeWhisper/Personal/PersonalSettingsSection.swift",
  "TypeWhisper/Personal/ReviewLearning.swift",
  "TypeWhisper/Personal/TranscriptReviewView.swift",
  "TypeWhisper/Personal/TranscriptReviewController.swift",
]

test_files = [
  "TypeWhisperTests/FuzzyTermMatcherTests.swift",
  "TypeWhisperTests/DictionarySyncMergeTests.swift",
  "TypeWhisperTests/ReviewLearningTests.swift",
]

def find_or_create_group(project, segments)
  group = project.main_group
  segments.each do |seg|
    existing = group.children.find { |c| c.is_a?(Xcodeproj::Project::Object::PBXGroup) && c.display_name == seg }
    group = existing || group.new_group(seg, seg)
  end
  group
end

def already_referenced?(project, path)
  project.files.any? { |f| f.real_path.to_s.end_with?(path) }
end

added = 0

app_files.each do |path|
  if already_referenced?(project, path)
    puts "skip (exists): #{path}"
    next
  end
  segments = File.dirname(path).split("/")
  group = find_or_create_group(project, segments)
  ref = group.new_reference(File.basename(path))
  main_target.add_file_references([ref])
  added += 1
  puts "added (app): #{path}"
end

test_files.each do |path|
  if already_referenced?(project, path)
    puts "skip (exists): #{path}"
    next
  end
  segments = File.dirname(path).split("/")
  group = find_or_create_group(project, segments)
  ref = group.new_reference(File.basename(path))
  test_target.add_file_references([ref])
  added += 1
  puts "added (test): #{path}"
end

project.save
puts "Done. #{added} file(s) added."
